pyiceberg/table/__init__.py

```diff
@@ -219,68 +220,41 @@ def property_as_int(properties: Dict[str, str], property_name: str, default: Opt
 
 class Transaction:
     _table: Table
+    _table_metadata: TableMetadata
+    _autocommit: bool
     _updates: Tuple[TableUpdate, ...]
     _requirements: Tuple[TableRequirement, ...]
 
-    def __init__(
-        self,
-        table: Table,
-        actions: Optional[Tuple[TableUpdate, ...]] = None,
-        requirements: Optional[Tuple[TableRequirement, ...]] = None,
-    ):
+    def __init__(self, table: Table, autocommit: bool = False):
         self._table = table
-        self._updates = actions or ()
-        self._requirements = requirements or ()
+        self._table_metadata = table.metadata
+        self._autocommit = autocommit
+        self._updates = ()
+        self._requirements = ()
 
     def __enter__(self) -> Transaction:
         """Start a transaction to update the table."""
         return self
 
     def __exit__(self, _: Any, value: Any, traceback: Any) -> None:
         """Close and commit the transaction."""
-        fresh_table = self.commit_transaction()
-        # Update the new data in place
-        self._table.metadata = fresh_table.metadata
-        self._table.metadata_location = fresh_table.metadata_location
-
-    def _append_updates(self, *new_updates: TableUpdate) -> Transaction:
-        """Append updates to the set of staged updates.
-
-        Args:
-            *new_updates: Any new updates.
+        self.commit_transaction()
 
-        Raises:
-            ValueError: When the type of update is not unique.
+    def _apply(self, updates: Tuple[TableUpdate, ...], requirements: Tuple[TableRequirement, ...] = ()) -> Transaction:
+        """Check if the requirements are met, and applies the updates to the metadata."""
+        for requirement in requirements:
+            requirement.validate(self._table_metadata)
 
-        Returns:
-            Transaction object with the new updates appended.
-        """
-        for new_update in new_updates:
-            # explicitly get type of new_update as new_update is an instantiated class
-            type_new_update = type(new_update)
-            if any(isinstance(update, type_new_update) for update in self._updates):
-                raise ValueError(f"Updates in a single commit need to be unique, duplicate: {type_new_update}")
-        self._updates = self._updates + new_updates
-        return self
+        self._updates += updates
+        self._requirements += requirements
 
-    def _append_requirements(self, *new_requirements: TableRequirement) -> Transaction:
-        """Append requirements to the set of staged requirements.
+        self._table_metadata = update_table_metadata(self._table_metadata, updates)
```

Comment:
Now there will be 2 updated metadata after the transaction commits: `_table_metadata` in the transaction and the new metadata of the table. I think it may be good to add a test to ensure that the 2 metadata are the same after the transaction. This could help verify the correctness of the internal `_table_metadata` WDYT?

Reply:
Yes, I also looked into this, but this breaks some of the APIs. The commit accepts a `CommitTableRequest` instead of the transaction. For example, for the REST catalog, we don't apply the changes to the table metadata.

Reply:
Is it possible to check the metadata equality like:
```python
with tbl.transaction() as txn:
  ...
assert tbl.metadata == txn.metadata
```
so that it can help verify that transaction accumulates the metadata change in the same way as catalog does

Reply:
The problem is that in the `Transaction`, we call the `_commit` method on the table, and we don't carry through the metadata there. For the REST catalog, this is not relevant, for the catalogs where you have to construct the metadata yourself, it is. I like these additional checks, but I'm unsure if we want to change the API to pass this through.